tieredhashing/tiered_hashing.go

```diff
@@ -199,43 +200,40 @@ func (t *TieredHashing) GetPoolMetrics() PoolMetrics {
 	}
 }
 
-func (t *TieredHashing) GetNodes(key string, n int) []string {
-	// TODO Replace this with mirroring once we have some metrics and correctness info
-	// Pick nodes from the unknown tier 1 in every 5 times
+func (t *TieredHashing) GetNodes(from Tier, key string, n int) []string {
 	var nodes []string
 	var ok bool
 
-	if !t.cfg.AlwaysMainFirst && rand.Float64() <= 0.2 {
-		if t.unknownSet.Size() != 0 {
-			nodes, ok = t.unknownSet.GetNodes(key, t.unknownPossible(n))
-			if !ok {
-				return nil
-			}
-			if len(nodes) == n {
-				return nodes
-			}
+	if from == TierUnknown {
+		nodes, ok = t.unknownSet.GetNodes(key, t.unknownPossible(n))
+		if !ok {
+			return nil
 		}
-
-		nodes2, _ := t.mainSet.GetNodes(key, t.mainPossible(n-len(nodes)))
-		nodes = append(nodes, nodes2...)
-	} else {
-		if t.mainSet.Size() != 0 {
-			nodes, ok = t.mainSet.GetNodes(key, t.mainPossible(n))
-			if !ok {
-				return nil
-			}
-			if len(nodes) == n {
-				return nodes
-			}
+		if len(nodes) == n {
+			return nodes
+		}
+	} else if from == TierMain {
+		nodes, ok = t.mainSet.GetNodes(key, t.mainPossible(n))
+		if !ok {
+			return nil
+		}
+		if len(nodes) == n {
+			return nodes
 		}
-
-		nodes2, _ := t.unknownSet.GetNodes(key, t.unknownPossible(n-len(nodes)))
-		nodes = append(nodes, nodes2...)
 	}
 
 	return nodes
 }
 
+func (t *TieredHashing) NodeTier(node string) Tier {
+	for k, n := range t.nodes {
+		if k == node {
```

Comment:
Let's copy `n` which is a pointer var before using it ?

Reply:
we're returning a direct string, so nothing about n is retained to get messed up outside of the loop